### ISSUE TITLE
Proposal: make language_version more intuitive or create minimum_language_version

### ISSUE THREAD

**Issue by rafrafek** [pre-commit/pre-commit#2640]
Please consider the situation below:

Hook creator knows that the hook cannot run on Python 2. The creator adds `language_version: python3` to the hook definition.

User has two Python 3 versions installed: Python 3.9 and Python 3.10. Python 3.9 is the system default, but the user wants to run all hooks with Python 3.10. The user specifies `default_language_version` to python3.10.

Hook runs on python3.9, because currently `language_version` causes `default_language_version` to be completely ignored.

## Proposal:

If the `default_language_version` matches `language_version`, use the one that is more specific. In the case above it would be `default_language_version`.

### Example:

`language_version` set to python3

`default_language_version` set to python3.10

Expected result:

Run hook with python3.10 because `default_language_version` matches `language_version`. Do not run with python3.9 or any other version that does not match both `language_version` and `default_language_version`.

### Example:

`language_version` set to python3.10

`default_language_version` set to python3.10.2

Expected result:

Run hook with python3.10.2 because `default_language_version` matches `language_version`. Do not run with python3.10.1 or any other version that does not match both `language_version` and `default_language_version`.

### Example:

`language_version` set to python3.7

`default_language_version` set to python3.10

Expected result:

Run hook with python3.7 because `default_language_version` does not match `language_version`.

### Another way that this issue could be solved:

Create `minimum_language_version` parameter that will indicate minimum language version and let hook creators use it instead of `language_version` for such cases.

**Comment by rafrafek** [comment]
Thank you for the link. I can see it is not implemented because of `virtualenv`'s lack of better Python version selection. 

Can it be implemented on the pre-commit side? A parser that compares versions before passing them to the `virtualenv`?

(Optional) Do you know if there is already an issue for that in the `virtualenv`'s repo? I'm trying to find something related in their issue tracker but with no success so far.

**Comment by asottile** [comment]
please read what I said in the other thread for my stance on the topic -- it's exhausting to have to re-litigate every decision -- please read what's already there

**Comment by rafrafek** [comment]
I don't want to waste your time, feel free to unsubscribe without replying. 

I hope it can be answered by the community.

In #1128 I can read that:
> there's nothing really that can be done

I couldn't find any explanation why. If anyone knows what is the main blocker for implementing it as a parser on the pre-commit side, please let me know.

There is already minimum_pre_commit_version implemented that uses simple parse_version parser from util.py and there is a version comparison for pre-commit version in repository.py. I don't see why it couldn't be implemented for language_version and default_language_version. 

The algorithm could look like this:

If there are both values set, check if default_language_version matches language_version and if yes, select version that is more specific, which basically means it has more dots.

**Comment by asottile** [comment]
version parsing for _pre-commit_ versions is trivial because I control the version string

version parsing for version strings which are passed to arbitrary third party tools is intractable and trying to do so would be error prone, complicated, and confusing to end users

**Comment by rafrafek** [comment]
Thank you! Now I can understand the downsides.

One last thing:

How about creating a new parameter that would work globally like the default_language_version, but also having higher precedence over language_version? It could be named force_language_version. 

With such parameter I could specify Python version in one line in my .pre-commit-config.yaml for all the hooks I'm using and be sure all of them use the version I want. That would be nice!

**Comment by rafrafek** [comment]
Yes, it would require two new force parameters, one for global scope and one for local.

From strongest to weakest:

current hierarchy:
1. language_version (local)
2. default_language_version (global)

new hierarchy:
1. force_language_version (local)
2. force_language_version (global)
3. language_version (local)
4. default_language_version (global)

or:
1. ignore_force_language_version (local) (flag)
2. force_language_version (global)
3. language_version (local)
4. default_language_version (global)

It will solve the problem but introduce some (optional) complexity.

---

The other option would be not solving the problem in pre-commit code but rather inform hook creators via the docs to avoid language_version when it is not strictly necessary to run a hook with a very specific version. This is at least valid for Python, where many hook creators just want to specify minimum Python version or specify generic python3 to avoid Python 2. And many users want all hooks to use the same Python version as their Python code.

---

Some notes

Current state problems from the user perspective:
- user may not know that some hooks use generic python3 version and may unconsciously run some hooks with a different language version than specified by default_language_version, which is not intended both by the user and the hooks creators
- user has to manually check all hooks definitions for existence of language_version
- user has to specify language_version for each hook with generic python3 version (breaks DRY)

Current state problems from the hook creator perspective:
- there is no way to specify minimum language version or supported versions range

**Comment by asottile** [comment]
yeah again, not happening as decided before. please stop re litigating prior decisions
